crates/sui-core/src/authority/authority_store.rs

```diff
@@ -608,43 +696,73 @@ impl AuthorityStore {
                 }
             })
             .collect()
     }
 
     /// Checks if the input object identified by the InputKey exists, with support for non-system
-    /// packages i.e. when version is None.
-    pub fn multi_input_objects_exist(
+    /// packages i.e. when version is None. If the input object doesn't exist and it's a receiving
+    /// object, we also check if the object exists in the object marker table and view it as
+    /// existing if it is in the table.
+    pub fn multi_input_objects_available(
         &self,
         keys: impl Iterator<Item = InputKey> + Clone,
+        receiving_objects: HashSet<InputKey>,
+        epoch_store: &AuthorityPerEpochStore,
     ) -> Result<Vec<bool>, SuiError> {
-        let (keys_with_version, keys_without_version): (Vec<_>, Vec<_>) =
-            keys.enumerate().partition(|(_, key)| key.1.is_some());
+        let (keys_with_version, keys_without_version): (Vec<_>, Vec<_>) = keys
+            .enumerate()
+            .partition(|(_, key)| key.version().is_some());
 
-        let versioned_results = keys_with_version.iter().map(|(idx, _)| *idx).zip(
+        let mut versioned_results = vec![];
+        for ((idx, input_key), has_key) in keys_with_version.iter().zip(
             self.perpetual_tables
                 .objects
-                .multi_get(
+                .multi_contains_keys(
                     keys_with_version
                         .iter()
-                        .map(|(_, k)| ObjectKey(k.0, k.1.unwrap())),
+                        .map(|(_, k)| ObjectKey(k.id(), k.version().unwrap())),
                 )?
-                .into_iter()
-                .map(|o| o.is_some()),
-        );
+                .into_iter(),
+        ) {
+            if has_key {
+                versioned_results.push((*idx, true))
+            } else if receiving_objects.contains(input_key) {
+                // For any objects that are unable to be fetched that are receiving objects, lookup
+                // and determine if the object exists in the object marker table as well. If so we
+                // will then mark it as "available" to let it progress through. If a version of the
+                // object greater than this one exists, then it was mutated by someone else and
+                // then deleted, and we can let it through to let the transaction fail at execution.
+                //
+                // If there's a more recent version of this object, then it was mutated by someone
+                // else and we can let it through to let the transaction fail at execution.
+                let has_received =
+                    self.have_received_deleted_or_wrapped_object_at_version_or_after(
+                        &input_key.id(),
+                        input_key.version().unwrap(),
+                        epoch_store.epoch(),
+                    )? || self
+                        .get_object(&input_key.id())?
```

Comment:
Good question. In the TM all checks related to receiving arguments -- both for existing objects and deletion markers -- have to be range based due to the fact that the object could be mutated by an intervening transaction and the old version pruned. So in particular we need to ask "does there exist an object version >= the version in the object ref" or "has this been deleted at a version >= the version in the object ref."

In particular, here we need to call `get_object` and cannot use `get_object_by_key` since the object could have been mutated but not deleted by an intervening transaction, and the old object version could have been pruned. In this case the object or versions great than it won't be in the marker table (since it wasn't deleted), and the object key won't exist either. But a more recent (active) version of the object does exist and this is where the `get_object` is important.